Mathlib/Logic/Function/OfHArity.lean

```diff
@@ -0,0 +1,86 @@
+/-
+Copyright (c) 2024 Brendan Murphy. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Brendan Murphy
+-/
+
+import Mathlib.Data.Fin.VecNotation
+
+/-! # Function types of a given heterogeneous arity
+
+This provides `FunctionOfHArity`, such that `OfHArity ![α, β] τ = α → β → τ`.
+Note that it is often preferable to use `((i : Fin n) → p i) → τ` in place of `OfHArity p τ`.
+
+## Main definitions
+
+* `Function.OfHArity p τ`: `n`-ary function `p 0 → p 1 → ... → p (n - 1) → β`.
+-/
+
+universe u
+
+namespace Function
+
+open Matrix (vecCons vecHead vecTail vecEmpty)
+
+/-- The type of `n`-ary functions `p 0 → p 1 → ... → p (n - 1) → τ`. -/
+def OfHArity : {n : ℕ} → (Fin n → Type u) → Type u → Type u
```

Comment:
I think this should perhaps be called `OfTypeTuple` or something? `OfArity` is so named because `OfArity A B 3` is a function of arity `3`. This declaration reads as `OfHArity ![B, C] D`, which doesn't actually mention the arity at all.

Reply:
I guess I what I had in mind is that `![B, C]` is the arity of the function, eg in the sense that a multi-sorted theory has function symbols whose arities look like `![A1,...,An] -> B`. But that interpretation is pretty unclear from just the code/docs, and not that standard to begin with. The name `OfTypeTuple` also feels unclear to me, what about `FromTypes ![B, C] D`?

Reply:
`FromTypes` sounds fine too. Maybe ask on Zulip for a name suggestion?